torch/csrc/distributed/c10d/socket.cpp

```diff
@@ -0,0 +1,818 @@
+  using Handle = SOCKET;
+#else
+  using Handle = int;
+#endif
+
+#ifdef _WIN32
+  static constexpr Handle invalid_socket = INVALID_SOCKET;
+#else
+  static constexpr Handle invalid_socket = -1;
+#endif
+
+  explicit SocketImpl(Handle hnd) noexcept
+      : hnd_{hnd} {}
+
+  SocketImpl(const SocketImpl& other) = delete;
+
+  SocketImpl& operator=(const SocketImpl& other) = delete;
+
+  SocketImpl(SocketImpl&& other) noexcept = delete;
+
+  SocketImpl& operator=(SocketImpl&& other) noexcept = delete;
+
+  ~SocketImpl();
+
+  std::unique_ptr<SocketImpl> accept() const;
+
+  void closeOnExec() noexcept;
+
+  void enableNonBlocking();
+
+  void disableNonBlocking();
+
+  bool enableNoDelay() noexcept;
+
+  bool enableDualStack() noexcept;
+
+#ifdef _WIN32
+  bool enableExclusiveAddressUse() noexcept;
+#else
+  bool enableAddressReuse() noexcept;
+#endif
+
+  std::uint16_t getPort() const;
+
+  Handle handle() const noexcept {
+    return hnd_;
+  }
+
+ private:
+  bool setSocketFlag(int level, int optname, bool value) noexcept;
+
+  Handle hnd_;
+};
+
+SocketImpl::~SocketImpl() {
+#ifdef _WIN32
+  ::closesocket(hnd_);
+#else
+  ::close(hnd_);
+#endif
+}
+
+std::unique_ptr<SocketImpl> SocketImpl::accept() const {
+  ::sockaddr_storage addr_s{};
+
+  auto addr_ptr = reinterpret_cast<::sockaddr*>(&addr_s);
+
+  ::socklen_t addr_len = sizeof(addr_s);
+
+  Handle hnd = ::accept(hnd_, addr_ptr, &addr_len);
+  if (hnd == invalid_socket) {
+    std::error_code err = getSocketError();
+    if (err == std::errc::interrupted) {
+      throw std::system_error{err};
+    }
+
+    std::string msg{};
+    if (err == std::errc::invalid_argument) {
+      msg = fmt::format("The server socket on {} is not listening for connections.", *this);
+    } else {
+      msg = fmt::format("The server socket on {} has failed to accept a connection {}.", *this, err);
+    }
+
+    logError(msg);
+
+    throw SocketError{msg};
+  }
+
+  ::addrinfo addr{};
+  addr.ai_addr = addr_ptr;
+  addr.ai_addrlen = addr_len;
+
+  logInfo("The server socket on {} has accepted a connection from {}.", *this, addr);
+
+  auto impl = std::make_unique<SocketImpl>(hnd);
+
+  // Make sure that we do not "leak" our file descriptors to child processes.
+  impl->closeOnExec();
+
+  if (!impl->enableNoDelay()) {
+    logWarning("The no-delay option cannot be enabled for the client socket on {}.", addr);
+  }
+
+  return impl;
+}
+
+void SocketImpl::closeOnExec() noexcept {
+#ifndef _WIN32
+  ::fcntl(hnd_, F_SETFD, FD_CLOEXEC);
+#endif
+}
+
+void SocketImpl::enableNonBlocking() {
+#ifdef _WIN32
+  unsigned long value = 1;
+  if (::ioctlsocket(hnd_, FIONBIO, &value) == 0) {
+    return;
+  }
+#else
+  int flg = ::fcntl(hnd_, F_GETFL);
+  if (flg != -1) {
+    if (::fcntl(hnd_, F_SETFL, flg | O_NONBLOCK) == 0) {
+      return;
+    }
+  }
+#endif
+  throw SocketError{"The socket cannot be switched to non-blocking mode."};
+}
+
+// TODO: Remove once we migrate everything to non-blocking mode.
```

Comment:
if this won't happen as an immediate followup PR, shall we create an issue to track this?

Reply:
Hope to continue working on this project and finish it before the end of the year. I think it makes sense to have a task anyways though for keeping record things.